Implement ES2018 Async Iteration (Experimental)

rhuanjl opened the pull request:
**This PR implements ES2018 Async Iteration (Async Generator Functions and for await):**
**UPDATE: This is now a complete implementation except:**
1. It needs more tests (help appreciated here)
2. I haven't looked at profiler support
3. I haven't looked at TTD support
4. This is placed behind a flag, ES2018AsyncIteration set to default to disabled  (hoping it can be enabled soon but feels too big to enable straight away)

**Implementation notes:**
1. AsyncGenerators are implemented by extending the existing Generator type
2. Handling await, yield and yield* in async generators has involved adding 4 new op codes - wanted it only to be 3 but couldn't find a way to handle the interaction between .return() and yield* without an extra op
3. AsyncFromSyncIterator (required per spec for async iteration of a synchronous object) has been added as a whole new type

**Test262 results**
1. I've run this against the test 262 tests in the following folders, I used  the ES6FunctionNameFull and ES2018ObjectRestSpread flags when running the tests as some of these tests require those features:
https://github.com/tc39/test262/tree/master/test/built-ins/AsyncFromSyncIteratorPrototype
https://github.com/tc39/test262/tree/master/test/built-ins/AsyncGeneratorFunction
https://github.com/tc39/test262/tree/master/test/built-ins/AsyncIteratorPrototype/Symbol.asyncIterator
https://github.com/tc39/test262/tree/master/test/language/statements/async-generator
https://github.com/tc39/test262/tree/master/test/language/statements/for-await-of

2. Most tests pass, fails relate to existing issues:
    a. Next isn't cached - issue exists with synch iterators https://github.com/Microsoft/ChakraCore/issues/5860
    b. the `async` keyword is not allowed to contain unicode escapes (affects normal async functions as well as async generators)
    c. `of` in a loop header is not allowed to contain a unicode escape
    d. 12 other issues to do with destructuring syntax and loop headers


ref: #2720

pleath commented:
Thanks, @rhuanjl . You mention that you're adding runtime checks. Do those checks determine which context await/yield are being employed in, and/or to distinguish between await and yield?

rhuanjl commented:
The runtime checks this contains which could clearly be removed by refactoring (likely with more code in total) are:
1. await vs yield vs yield*
2. AsyncGenerator vs Generator (when calling the next/throw/return methods have to check type twice as first check gives same answer for both as it's the same class)

pleath commented:
I see. Yes, you're right, we want to be able to resolve static properties of the code without using runtime checks. A separate opcode for yield in different contexts is a good place to start and something we anticipated having to do to implement the feature. I'll be glad to help you do that. We can handle the JIT after getting things working in the interpreter. You'll need to:
- add your new opcode to runtime/bytecode/opcodes.h
- make the byte code generator emit your opcode in the proper context (runtime/bytecode/bytecodeemitter.cpp)
- add a handler for your opcode to the interpreter, which will likely just route to a runtime helper.

rhuanjl commented:
Whilst this version should work as is certainly agree that it could do with improvements, I'll aim to submit an update with additional opcodes by the end of this week or start of next.

rhuanjl commented:
I don't understand the assertion this hits - my guess is it is to do with my new OpCodes either:
a) Using the ResumeYieldData from the stack OR
b) Using the RHS value without copying it to a temporary register

It runs fine in a test build so is it just an attribute I need to set somewhere to avoid the assertion?

**Changes made:**
I've made the following changes from the original version:
1. fixed some errors I'd made around incorrect handling of returns
1. Introduced OpCodes for Await, AsyncYield and AsyncYield*
1. Implemented the OpCodes as runtime function calls
1. added the OpCodes as helper calls in the Jit (I think I have this right...) 
1. Removed most of the temporary objects my initial version needed (including pre-creating and holding onto continuation functions for await and yield as with most async generators these will likely be needed multiple times)
1. Significantly reduced the number of runtime checks/branches required
1. Updated my handling of await for a normative PR that's currently open to the spec (it already has consensus so should in theory be being merged soon)

**OtherNotes**
1. Hopefully the assertion I'm hitting is an easy fix though I'd appreciate help with that - I've had a look but can't figure it out.
1. If/when this is merged I'd like to follow up with an edit to the Async function implementation to use much of this code - will be able to reduce a lot of duplication as well as streamlining async functions
1. Pending points from my first comment remain the same

rhuanjl commented:
Small update - I'm about half way there offline with the remaining pieces needed to meet the spec i.e. for-await-of and async from sync iterator. It's going to be 500+ more lines of code when done so may be easier if this part is reviewed first before I add the rest?

pleath commented:
Thanks, @rhuanjl . Are you still hitting the assert? If so, can you give me some detail about it?

fatcerberus commented:
@rhuanjl I mentioned this in private already but it seems that this feature currently can be enabled only with a flags-enabled build of `ch` (using the command-line option).  JsRuntimeAttributeEnableExperimentalFeatures doesn’t enable it so there’s no way to use the feature at all in a Release build.

rhuanjl commented:
I'm afraid my yield* implementation is wrong both here and in the further work I've done offline - it's right 95% of the time but gets edge cases wrong - struggling to come up with a way to handle it correctly and nicely other than the option of not doing the yield* loop in bytecode.

As far as I can see this would be:
- much easier to implement the spec
- probably more performant (as yield* would no longer involve yield/resume for each iteration)

BUT for error handling the stack trace if the iterator threw would not include the generator function which would not be ideal.

rhuanjl commented:
UPDATE: see edited first post for details - this is now a complete implementation of ES2018 Async Iteration (though could do with more tests)

Please could people find some time to begin reviewing this?

pleath commented:
Awesome. Thanks, @rhuanjl . I'll give this a close review this week.

zenparsing commented:
Thanks @rhuanjl and @pleath - it will be really awesome to get this in!

pleath commented:
FYI, I'm about half way through the review, and so far everything looks nice and clean. I want to walk through a couple of use cases before I sign off.

fatcerberus commented:
This still is not usable with `JsRuntimeAttributeEnableExperimentalFeatures`, which is the only way to get access to experimental stuff in the release build - can that be fixed?

rhuanjl commented:
@pleath thanks for the prelim feedback - I've done a re-read/check and noticed a couple of things, not going to update the PR whilst you're mid review, notes for ref:
1. I realised that the JIT handlers I included in lower.cpp are untested as by default jilting generators is disabled (controlled by flag JitES6Generators) BUT - it's disabled as it doesn't work - jitting sync generators causes a crash - async ones crash the same way - so don't know if my code is ok and it's just the existing problem that's an issue or if my code is wrong.

1. Some of my spec text comments have the wrong numbering (oops, I'll fix this in a separate commit after you've reviewed)

1. to @fatcerberus point on JsRuntimeAttributeEnableExperimentalFeatures it seems this enables flags that use the macro FLAGPR_REGOVR_EXP() and not FLAGPR() - not sure which this should be

1. Not sure what to do for better test coverage - the basic cases I've included clearly aren't enough - test262 has ~2000 tests for this; it might take a long time to create full coverage.

rhuanjl commented:
@pleath Any update on this - as fixing the hit piece is out of scope per discussion in #5877 only other points I'm aware of are that it probably needs more tests and a couple of my comments are wrong - didn't want to add anything when you were mid-review though

(Also really expected your review to find some errors in my code this is the largest piece of c++ I've ever written)

pleath commented:
(Hah. My previous comment was mangled, but I’ve fixed it.)

pleath commented on lib/Runtime/Library/JavascriptGenerator.cpp:
```diff
@@ -247,6 +262,419 @@ namespace Js
+        Var result = nullptr;
+        JavascriptExceptionObject *exception = nullptr;
+        yieldData->generator = this;
+        AssertMsg(isAsync, "Should not call CallAsyncGenerator on a non-async generator");
```
Can this be AssertOrFailFast instead, just to forestall a type confusion security issue if a mistake is made in a caller somewhere down the road?

rhuanjl replied:
In most cases the caller checks this so making it a FailFast means it's checked twice - but agree that getting this wrong would be an easy way to introduce a security bug in the future. Will update.

pleath commented on lib/Runtime/Library/JavascriptGenerator.cpp:
```diff
@@ -247,6 +262,419 @@ namespace Js
+            public:
+                GeneratorStateHelper(JavascriptGenerator* g) : g(g), didThrow(true) { g->SetState(GeneratorState::Executing); }
+                ~GeneratorStateHelper() { if (didThrow || g->frame == nullptr) {g->SetState(GeneratorState::Completed); } }
+                void DidNotThrow() { didThrow = false; }
```
Hm. I'm not sure we can ever execute a generator without having created an interpreter frame instance. One design issue in the JIT scenario that we haven't fully worked out is the preservation of state between a yield and a next. We may need to have an interpreter instance to support this even after the generator has been jitted. For now, I would assume that an interpreter frame is always present.

rhuanjl replied:
In that case I'll see if I can re-work the logic here a bit to simplify it - likely doing some unnecessary work.

pleath commented on lib/Runtime/Library/JavascriptGenerator.cpp:
```diff
@@ -247,6 +262,419 @@ namespace Js
+    {
+        JavascriptPromise* promise = JavascriptPromise::CreateEnginePromise(scriptContext);
+
+        if (!VarIs<JavascriptGenerator>(thisValue))
```
Is it possible to hit this path without an internal error of some kind, and is it a case where the spec calls for a type error?

rhuanjl replied:
Yes this would be if you did something like:
```js
async function* asyncGenF() {}
const nextMethod = asyncGenF().next;
nextMethod.call(anything);
```
Relevant spec text is here - it does have a type error for this:https://tc39.github.io/ecma262/#sec-asyncgeneratorenqueue

I'll add the spec text as comments to this function - not sure why I didn't add them before.

pleath commented on lib/Runtime/Library/JavascriptGenerator.cpp:
```diff
@@ -247,6 +262,419 @@ namespace Js
+
+        JavascriptGenerator* generator = UnsafeVarTo<JavascriptGenerator>(thisValue);
+
+        if (!generator->GetIsAsync())
```
(Same question as above.)

rhuanjl replied:
Yes similar to point above though this one would be specifically if you did:
```js
function* genF(){}
async function* asyncGenF(){}
const nextMethod = asyncGenF().next;
nextMethod.call(genF());
```

rhuanjl commented:
1. I've squashed my previous commits
2. Rebased to latest master and regen'd byte code (to fix merge conflict)
3. Updated the comment error + points from @pleath's review as the new commit "fixes"

Note this PR still adds this as a disabled feature - partly as no one has told me to enable it since I opened the PR and partly as I'm worried my test coverage is insufficient for a shipping feature this big.

fatcerberus commented:
Keep in mind that if the feature is fully disabled, including with the EnableExperimentalFeatures runtime attribute, then nobody can test it unless they have a flags-enabled build of CC.  That's not likely to be the case for embedders shipping an application, so the feature will get no real-world testing.

rhuanjl commented:
@fatcerberus I'm hoping this can be enabled soon but leaving that decision to the CC team - as an additional note node-chrakracore uses EnableExperimentalFeatures so if I switched it to an experimental feature of that kind it would be on in node-chakracore - it should really have TTD support before then I assume; and I don't know how to do TTD support.

fatcerberus commented:
> I'm hoping this can be enabled soon but leaving that decision to the CC team

Yes, I was aware - I was simply bringing the above up as a discussion point.

rhuanjl commented:
Would it be possible to move forward with this or is it parked indefinitely?

Is there anything specific I can do to move this towards being mergeable?

zenparsing commented:
Hi @rhuanjl - Thanks for pinging on this! I intend to pick this up within the next few days. In the meantime, would you mind doing a rebase?

rhuanjl commented:
@zenparsing I saw it got out of date, and had bytecode conflicts so I've just regenerated the bytecode and rebased again. I also squashed the fixes commit for neatness.

Please let me know if you'd like me to do anything else on this.

bathos commented on test/es7/async-generator-apis.js:
```diff
@@ -0,0 +1,216 @@
+        }
+    },
+    {
+        name : "Async Generaotr function instances have the correct prototype object",
```
```suggestion
        name : "Async Generator function instances have the correct prototype object",
```

rhuanjl replied:
I've fixed this typo in the follow up PR #6172 thanks.

rhuanjl commented:
Thanks for merging this @zenparsing 

As noted above whilst this is a full implementation it's disabled by default - please let me know if there's anything I contribute to help get towards it being enabled by default.
